pairtools/lib/fileio.py

```diff
@@ -1,156 +1,144 @@
 import shutil
-import pipes
 import subprocess
+import shlex
 import sys
+import typing as tp
+from dataclasses import dataclass
 
 class ParseError(Exception):
     pass
 
 
-def auto_open(path, mode, nproc=1, command=None):
-    """Guess the file format from the extension and use the corresponding binary
-    to open it for reading or writing. If the extension is not known, open the
-    file as text.
-
-    If the binary allows parallel execution, specify the number of threads
-    with `nproc`.
-
-    If `command` is supplied, use it to open the file instead of auto-guessing.
-    The command must accept the filename as the last argument, accept input
-    through stdin and print output into stdout.
-
-    Supported extensions and binaries (with comments):
-    .bam - samtools view (allows parallel writing)
-    .gz - pbgzip if available, otherwise bgzip
-    .lz4 - lz4c (does not support parallel execution)
-    """
-
-    # Empty filepath or False provided
-    if not path or path == "-":
-        if mode == "r":
-            return sys.stdin
-        if mode == "w":
-            return sys.stdout
-
-    if command:
-        if mode == "w":
-            t = pipes.Template()
-            t.append(command, "--")
-            f = t.open(path, "w")
-        elif mode == "r":
-            t = pipes.Template()
-            t.append(command, "--")
-            f = t.open(path, "r")
+MODES_TO_FILES_PRESET = {
+    'bam': ['w', 'r'],
+    'gz': ['w', 'r', 'a'],
+    'lz4': ['w', 'r', 'a'],
+}
+
+COMMANDS = {
```

Comment:
Would be great to write down the logic in the comment here: the first command is the first choice, etc.